### PR TITLE
Rent byte buffers in SqlSequentialTextReader

### PR DESCRIPTION
**Pull request by TrayanZapryanov**
While profiling our code I saw a lot of allocations when using reader.GetTextReader() over big string column when reader is opened with CommandBehavior.SequentialAccess.

image

This PR tries to fix this by renting byte[] from shared pool.

### CODE REVIEW

**Comment by JRahnama** [comment]
failures seems related to changes:
```ps
[xUnit.net 00:01:02.24]     Microsoft.Data.SqlClient.ManualTesting.Tests.DataStreamTest.RunAllTestsForSingleServer_NP [FAIL]
  [xUnit.net 00:01:02.25]       System.ArgumentException : Offset and length were out of bounds for the array or count is greater than the number of elements from index to the end of the source collection.
  [xUnit.net 00:01:02.25]       Stack Trace:
  [xUnit.net 00:01:02.25]            at System.Buffer.BlockCopy(Array src, Int32 srcOffset, Array dst, Int32 dstOffset, Int32 count)
  [xUnit.net 00:01:02.25]         /_/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlSequentialTextReader.cs(410,0): at Microsoft.Data.SqlClient.SqlSequentialTextReader.DecodeBytesToChars(Byte[] inBuffer, Int32 inBufferCount, Char[] outBuffer, Int32 outBufferOffset, Int32 outBufferCount)
  [xUnit.net 00:01:02.25]         /_/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlSequentialTextReader.cs(323,0): at Microsoft.Data.SqlClient.SqlSequentialTextReader.InternalRead(Char[] buffer, Int32 index, Int32 count)
  [xUnit.net 00:01:02.25]         /_/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlSequentialTextReader.cs(88,0): at Microsoft.Data.SqlClient.SqlSequentialTextReader.Read()
  [xUnit.net 00:01:02.25]         /_/src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/DataStreamTest/DataStreamTest.cs(1699,0): at Microsoft.Data.SqlClient.ManualTesting.Tests.DataStreamTest.ReadTextReader(String connectionString)
  [xUnit.net 00:01:02.25]         /_/src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/DataStreamTest/DataStreamTest.cs(178,0): at Microsoft.Data.SqlClient.ManualTesting.Tests.DataStreamTest.RunAllTestsForSingleServer(String connectionString, Boolean usingNamePipes)
  [xUnit.net 00:01:02.25]         /_/src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/DataStreamTest/DataStreamTest.cs(30,0): at Microsoft.Data.SqlClient.ManualTesting.Tests.DataStreamTest.RunAllTestsForSingleServer_NP()
```

**Comment by TrayanZapryanov** [comment]
Yes, I saw it and I have idea why. Rent method might return bigger array and calculations have to be changed to not use array limit. I will change it tomorrow.

**Comment by Wraith2** [comment]
The flow between the three methods is confusing but the code seems right overall. Can you remove the use of array.Empty? If you do that then you'll know that any non-null array is rented and can be returned, at the moment there could be way to try and return array.Empty to a pool which would either be bad or an exception.

In terms of style can you add braces to the single line conditionals please.

**Comment by TrayanZapryanov** [comment]
@Wraith2 I think empty array is used to progress reader only.
Check comment for [0 byte reads ](https://github.com/dotnet/SqlClient/blob/e3cf59d226675acf7ae9169989ac692aea04f3da/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlSequentialTextReader.cs#L172).
It doesn't look safe to remove if - just a check when returning should be enough to not give it back to pool.

**Comment by TrayanZapryanov** [comment]
@Wraith2 Ready for review. Build and tests are green

**Review by Wraith2 — APPROVED** [review]

**Comment by TrayanZapryanov** [comment]
@JRahnama Is this PR acceptable?

**Review by David-Engel — APPROVED** [review]
LGTM

**Comment by TrayanZapryanov** [comment]
When is expected this one to be merged?

**Comment by David-Engel** [comment]
> When is expected this one to be merged?

Just waiting on a second reviewer to pick it up then we'll merge it.

**Review by cheenamalhotra — APPROVED** [review]

**Pull request merged**
